Add support for Polls

DA-344 opened the pull request:
## Summary

<!-- What is this pull request for? Does it fix any issues? -->

Adds support for Polls.

### Related PRs:
- https://github.com/discord/discord-api-docs/pull/6746


### Additions:
- `Poll`, `PollAnswer` and `PollMedia` objects.
- `PollLayoutType` enum
- `on_poll_vote_add`/`remove` events.
- `poll` attribute to `Message`
- `poll` kwarg in `Messageable.send`

## Checklist

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

Gobot1234 commented on discord/poll.py:
```diff
@@ -0,0 +1,215 @@
+
+    @classmethod
+    def _from_data(cls, data: PollAnswerPayload) -> Self:
+        media = data.get('poll_media')
+        text = media.get('text')
+        emoji = media.get('emoji', None)
```
```suggestion
        media = data['poll_media']
```
if media is None other stuff breaks anyway.

DA-344 replied:
Yeah, you are right, I committed this change (with some others that are pretty similar like this).

No767 commented:
Is there documentation or an announcement surrounding this feature?

DA-344 commented:
> Is there documentation or an announcement surrounding this feature?

Not yet, but it is getting a pop-up saying you to try it, though you cannot send them (you get a 403: Forbidden).

Rapptz commented:
Thanks for the PR.

Discord has asked that I do not implement this yet and will give me private documentation when the time comes.

AbstractUmbra commented on discord/poll.py:
```diff
@@ -0,0 +1,644 @@
+
+        return cls(text=data['text'], emoji=PartialEmoji.from_dict(data.get('emoji', {})))
+
+    def __iter__(self) -> Union[str, Optional[PollMediaEmoji]]:
+        yield self.text
+        yield self.emoji
```
Why do we define `__iter__` here?

DA-344 replied:
Oh, I forgot to remove that! Sorry!

DA-344 replied:
This was removed as I said.

AbstractUmbra commented on discord/poll.py:
```diff
@@ -0,0 +1,644 @@
+
+
+
+class PollMedia:
```
I think this could be a NamedTuple subclass?

DA-344 replied:
I thought of it when I first created it, but I didn't subclass it at the end.
Going to change though.

DA-344 replied:
You are right, this is now a NamedTuple.

AbstractUmbra commented on discord/poll.py:
```diff
@@ -0,0 +1,644 @@
+    ----------
+    id: :class:`int`
+        The answer ID.
+    me_voted: :class:`bool`
```
I know this parameter name comes from the API payload but I personally find it the naming a bit lacking. How about `self_vote` or something along those lines?

DA-344 replied:
Yeah, will consider it 🤔

AbstractUmbra reviewed (commented):
Small items to check.

Also if this PR now "works as intended" as has been stated then please un-draft it so that CI can be run.

DA-344 commented:
I found an error when adding an emoji to a poll answer. Going to fix it ASAP.

DA-344 commented:
All errors I found with the PR have been fixed.

__Fixed errors__:
- `Message.poll` raising an error
- `PollAnswer.from_params` not allowing emojis
- `Poll._from_data` raising errors even if the duration exceeded some milliseconds

__Breaking changes__ *(for the ones that are using the branch)*:
*None*

__Changes__:
- Fixed type hinting errors
- Modified internal parameters (i.e.: PollAnswer)
- `Message.poll` default parameter value is now `None`, this has been done to prevent `AttributeError`s when using `poll.message` and when adding poll answers.

**If you see any errors please tell me**

bijij commented:
Out of curiosity, since the official docs don't seem to note it, can a poll be sent in a message with additional data such as attchements, embeds, content etc?

or is it some kind of mutually exclisive thing?

DA-344 commented:
> Out of curiosity, since the official docs don't seem to note it, can a poll be sent in a message with additional data such as attchements, embeds, content etc?

No, if you add any of them you get an error. It would be good if you could though.

bijij commented on discord/flags.py:
```diff
@@ -1255,6 +1255,55 @@ def auto_moderation_execution(self):
+
+    @flag_value
+    def guild_polls(self):
+        """:class:`bool`: Whether poll related events are enabled.
```
This should note this is specific to guilds.

bijij commented on discord/flags.py:
```diff
@@ -1255,6 +1255,55 @@ def auto_moderation_execution(self):
+
+    @flag_value
+    def dm_polls(self):
+        """:class:`bool`: Whether poll related events are enabled.
```
This should note this is specific to DMs.

bijij commented on discord/flags.py:
```diff
@@ -1255,6 +1255,55 @@ def auto_moderation_execution(self):
+    def dm_polls(self):
+        """:class:`bool`: Whether poll related events are enabled.
+
+        See also :attr:`guild_polls` for guilds or :attr:`polls` for both.
```
```suggestion
        See also :attr:`guild_polls` and :attr:`polls`.
```

bijij commented on discord/flags.py:
```diff
@@ -1255,6 +1255,55 @@ def auto_moderation_execution(self):
+    def guild_polls(self):
+        """:class:`bool`: Whether poll related events are enabled.
+
+        See also :attr:`dm_polls` for DMs or :attr:`polls` for both.
```
```suggestion
        See also :attr:`guild_polls` and :attr:`polls`.
```

bijij commented on discord/flags.py:
```diff
@@ -1255,6 +1255,55 @@ def auto_moderation_execution(self):
 
+    @flag_value
+    def polls(self):
+        """:class:`bool`: Whether poll related events are enabled.
```
Probably worth noting this is a combination of `guild_polls` and `dm_polls`

trevorflahardy commented on discord/poll.py:
```diff
@@ -0,0 +1,630 @@
+        layout_type = try_enum(PollLayoutType, data.get('layout_type', 1))
+        question_data = data.get('question')
+        question = question_data.get('text')
+        expiry = datetime.datetime.fromisoformat(data['expiry'])  # If obtained via API, then expiry is set.
```
`utils.parse_time` should be used here

trevorflahardy reviewed (commented):
Just some things I noticed whilst browsing. There are more, but they're more concept-related rather than strictly code edits. The fashion in which a `Poll` is both stateful and stateless will need to be addressed differently, as well as how `PollAnswer` is constructed and handled. 👍

trevorflahardy commented on discord/http.py:
```diff
@@ -154,6 +156,7 @@ def handle_message_parameters(
     thread_name: str = MISSING,
     channel_payload: Dict[str, Any] = MISSING,
     applied_tags: Optional[SnowflakeList] = MISSING,
+    poll: Optional[Poll] = MISSING
```
> i would disagree, since it makes pyright angry on `await ctx.send('Hello', poll=poll)` where `poll` is `Optional[Poll]`

Ah yes, you're right. Additionally, `abc.Messageable.send` should have `poll: Optional[Poll] = None` and not `poll: Poll = MISSING`, like it has now.

trevorflahardy replied:
> i would disagree, since it makes pyright angry on `await ctx.send('Hello', poll=poll)` where `poll` is `Optional[Poll]`

That would go back to the overload for `Context.send`. It's the same reason that you get an error when you do
![image](https://github.com/Rapptz/discord.py/assets/75498301/2b0c11fe-4e99-4364-b257-bf0ad9e4136b)

The revision suggestion only addressed `handle_message_parameters`, which denotes that `poll=None` cannot be given to either a `.send` or a `.edit` method.

trevorflahardy replied:
You're right, I was thrown off by the type hint for the send coroutines.

`handle_message_parameters` should have `Optional[Poll] = None` and all the send methods should have `poll: Optional[Poll] = None` in their function definition rather than `poll: Poll = MISSING`. In the `handle_message_parameters` block it should have `if poll is not None: ...`

```suggestion
    poll: Optional[Poll] = None
```
```python
def handle_message_parameters(poll: Optional[Poll] = None) -> ...:
    if poll is not None: ...
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,686 @@
+            # users is now a list of User
+
+        Parameters
+        ----------
+        limit: Optional[:class:`int`]
+            The max number of users to return. Can be up to 100.
+        after: Optional[:class:`abc.Snowflake`]
+            Fetches users after this ID, or every user if ``None``.
+
+        Raises
+        ------
+        HTTPException
+            Retrieving the users failed.
+
+        Yields
+        ------
+        Union[:class:`User`, :class:`Member`]
+            The users that voted for this poll answer. This can be a :class:`Member` object if the poll
+            is in a guild-message context, for other contexts it will always return a :class:`User`, or when
+            the member left the guild.
+        """
+
+        if not self._message or not self._state:  # Make type checker happy
+            raise RuntimeError('You cannot fetch users in a non-message-attached poll')
+
+        data = await self._state.http.get_poll_answer_voters(
+            self._message.channel.id, self._message.id, self.id, after.id if after else None, limit
+        )
+
+        if not self._message.guild:
+            for user in data.get('users'):
+                yield User(state=self._state, data=user)
+
+        else:
+            guild = self._message.guild
+
+            for user in data.get('users'):
+                member = guild.get_member(int(user['id']))
+
+                yield member or User(state=self._state, data=user)
```
This needs to be reworked entirely, if the limit is >100 we should instead keep hitting the endpoint until it is exhaused, reaction users does the same thing.

Additionally, I believe this should be called `voters`

DA-344 commented on discord/types/poll.py:
```diff
@@ -0,0 +1,95 @@
+
+
+class PollAnswerWithID(PollAnswer):
+    answer_id: Snowflake
```
No, thanks for reminding me of it! Fixed in latest commits.

DA-344 commented on discord/webhook/async_.py:
```diff
@@ -1732,6 +1749,15 @@ async def send(
 
             .. versionadded:: 2.4
 
+        poll: :class:`Poll`
```
No, I genuinely forgot about webhooks, thanks for reminding me lol

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,685 @@
+    @property
+    def original_message(self) -> Message:
+        """:class:`Message`: Returns the original message the poll of this answer is in."""
+        return self._message # type: ignore # Here will always be a value
```
```suggestion
        return self._message  # type: ignore # Here will always be a value
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,685 @@
+        poll_media: PollMediaPayload = {'text': text}
+        if emoji:
+            if isinstance(emoji, Emoji):
+                poll_media['emoji'] = {'name': emoji.name} # type: ignore
```
```suggestion
                poll_media['emoji'] = {'name': emoji.name}  # type: ignore
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,685 @@
+                poll_media['emoji'] = {'name': emoji.name} # type: ignore
+
+                if emoji.id:
+                    poll_media['emoji']['id'] = emoji.id # type: ignore
```
```suggestion
                    poll_media['emoji']['id'] = emoji.id  # type: ignore
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,685 @@
+            elif isinstance(emoji, PartialEmoji):
+                poll_media['emoji'] = emoji.to_dict()
+            else:
+                poll_media['emoji'] = {'name': str(emoji)} # type: ignore
```
```suggestion
                poll_media['emoji'] = {'name': str(emoji)}  # type: ignore
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,685 @@
+            if isinstance(self.emoji, PartialEmoji):
+                data['emoji'] = self.emoji.to_dict()
+            else:
+                data['emoji'] = {'name': str(self.emoji)} # type: ignore
```
```suggestion
                data['emoji'] = {'name': str(self.emoji)}  # type: ignore
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,685 @@
+                data['emoji'] = {'name': str(self.emoji)} # type: ignore
+
+                if hasattr(self.emoji, 'id'):
+                    data['emoji']['id'] = int(self.emoji.id) # type: ignore
```
```suggestion
                    data['emoji']['id'] = int(self.emoji.id)  # type: ignore
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,685 @@
+        self._message: Optional[Message] = None
+        self._state: Optional[ConnectionState] = None
+        self._finalized: bool = False
+        self._counts: Optional[Dict[int, PollAnswerCount]] = None # {answer_id: answer_count}
```
```suggestion
        self._counts: Optional[Dict[int, PollAnswerCount]] = None  # {answer_id: answer_count}
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,685 @@
+            answer._update(message)
+
+    @classmethod
+    def _from_data(
+        cls, *, data: PollPayload, message: Message, state: ConnectionState
+    ) -> Self:
```
```suggestion
    def _from_data(cls, *, data: PollPayload, message: Message, state: ConnectionState) -> Self:
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,685 @@
+        question_data = data.get('question')
+        question = question_data.get('text')
+        expiry = utils.parse_time(data['expiry'])  # If obtained via API, then expiry is set.
+        duration = expiry - message.created_at 
```
```suggestion
        duration = expiry - message.created_at
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,685 @@
+            'question': self._question_media.to_dict(),
+            'duration': self._hours_duration,
+            'layout_type': self.layout_type.value,
+            'answers': [
+                {'poll_media': answer._to_dict()}
+                for answer in self.answers
+            ]
```
```suggestion
            'answers': [{'poll_media': answer._to_dict()} for answer in self.answers],
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,685 @@
+        """
+
+        if self._message:
+            raise RuntimeError(
+                'Cannot append answers a poll recieved via a message'
+            )
```
```suggestion
            raise RuntimeError('Cannot append answers a poll recieved via a message')
```

bijij reviewed (changes requested):
I've gone through and marked where type-checkers or the formatter will fail this code.

We adhere strictly to our black formatting and pyright configuration.
I suggest you ensure in future you before committing run black and pyright over the `discord` directory in the repo.

bijij commented on discord/state.py:
```diff
@@ -1612,6 +1612,46 @@ def parse_entitlement_delete(self, data: gw.EntitlementDeleteEvent) -> None:
+            user = self.get_user(raw.user_id)
+
+        if message and user:
+            self.dispatch('poll_vote_add', user, message, message.poll.get_answer(raw.answer_id)) # type: ignore
```
Probably worth logging some kind of message in the event that we have the message object in cache but it doesn't have a poll. It's technically an error state.

DA-344 commented on discord/poll.py:
```diff
@@ -0,0 +1,630 @@
+        """:class:`str`: Returns this poll answer question string."""
+        return self._question_media.text
+
+    @property
+    def emoji(self) -> Optional[PartialEmoji]:
+        """Optional[:class:`PartialEmoji`]: Returns the emoji for this poll's question."""
+
+        return None  # As of now, polls questions don't support emojis
```
When Discord adds emoji support for questions this will be needed again, so it is better to return ``None`` until then.

bijij replied:
```suggestion
```
Disagree, when could be never.

bijij replied:
I'd rather this be left open for @Rapptz to chime in.

DA-344 commented on discord/raw_models.py:
```diff
@@ -503,3 +505,32 @@ def __init__(self, *, data: GuildApplicationCommandPermissions, state: Connectio
         self.permissions: List[AppCommandPermissions] = [
             AppCommandPermissions(data=perm, guild=self.guild, state=state) for perm in data['permissions']
         ]
+
+class RawPollVoteActionEvent(_RawReprMixin):
+    """Represents the payload for a :func:`on_raw_poll_vote_add` or :func:`on_raw_poll_vote_remove`
+    event.
+    
```
Changes made. (Forgot to add newline between classes)

bijij commented on discord/state.py:
```diff
@@ -1612,6 +1612,54 @@ def parse_entitlement_delete(self, data: gw.EntitlementDeleteEvent) -> None:
+
+        if message and user:
+            if not message.poll:
+                _log.warning('Cannot dispatch "poll_vote_add" event because the cached message does not contain a poll. This is a caching error.')
```
```suggestion
                _log.warning(
                    'POLL_VOTE_ADD referencing message with ID: %s without a poll. Discarding.',
                    raw.message_id,
                )
```

I'd probably reword this, same for the one for remove, at a mimimum including the message ID would be desirable.

bijij commented on discord/http.py:
```diff
@@ -256,6 +259,13 @@ def handle_message_parameters(
         }
         payload.update(channel_payload)
 
+    if poll is not MISSING:
+        if len(poll) == 0 or len(poll) > 10:
+            raise ValueError('Poll must contain between 1 and 10 answers')
+        if poll._hours_duration < 1 or poll._hours_duration > 168:
+            raise ValueError('Polls duration must be between 1 hour and 7 days')
+        payload['poll'] = poll._to_dict()  # type: ignore
```
You need to check poll isn't None prior to this.

DA-344 replied:
Fixed in latest commits

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,720 @@
+                yield member or User(state=self._state, data=raw_member)
+
+
+class PollAnswerCount(PollAnswerBase):
```
I feel like this functionality should be merged with PollAnswer and PollAnswer base shouldn't exist.

DA-344 replied:
As both poll answers allow this functionality, I thought it would be convinient to create a class where both PollAnswer & PollAnswerCount inherit from which already implements this method.

DA-344 replied:
You're right, I thought of it previously, but left it like that. I was already planning in merging them into ``PollAnswer``

bijij commented on discord/state.py:
```diff
@@ -1612,6 +1612,54 @@ def parse_entitlement_delete(self, data: gw.EntitlementDeleteEvent) -> None:
         entitlement = Entitlement(data=data, state=self)
         self.dispatch('entitlement_delete', entitlement)
 
+    def parse_message_poll_vote_add(self, data: gw.PollVoteActionEvent) -> None:
```
You should update the answer count, as well as the self_voted property when you recieve these events.

DA-344 commented on discord/poll.py:
```diff
@@ -0,0 +1,630 @@
+
+    .. container:: operations
+
+        .. describe:: str(x)
```
Fixed.

DA-344 commented on discord/poll.py:
```diff
@@ -0,0 +1,630 @@
+
+    Parameters
+    ----------
+    question: Union[:class:`PollMedia`, :class:`str`]
```
Using a ``str`` is a shortcut of ``PollMedia(text="...")``, and is internally handled.

DA-344 commented:
The PR is ready to be merged. Here's a summary of the changes:
- Added `on_poll_vote_add`/`on_poll_vote_remove` events. (With their respective raw events)
- New `Poll` class to support, well... polls. Can be subclassed.
- Implemented `PollAnswer` and `PollAnswerCount`
- Added `poll` parameter to ``Messageable.send``, ``PartialMessage.reply`` & ``Context.send`` (w/ ``Context.reply``)
- New ``end_poll`` method for ``PartialMessage``. Returns a ``Message`` on success.
- Added ``Permissions.send_polls``
- New ``RawPollVoteActionEvent`` raw model
- New ``Client.wait_for`` overloads to type hint poll related events.
- ``PollLayoutType`` enum.
- New ``Intents.polls``, ``Intents.guild_polls`` and ``Intents.dm_polls`` intents.

bijij commented on discord/message.py:
```diff
@@ -1937,6 +1969,11 @@ def _handle_components(self, data: List[ComponentPayload]) -> None:
     def _handle_interaction(self, data: MessageInteractionPayload):
         self.interaction = MessageInteraction(state=self._state, guild=self.guild, data=data)
 
+    def _handle_poll_vote(self, answer_id: int, added: bool):
```
Is there any reason for this function to exist?
Why can't the state just mess with the poll directly?

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,640 @@
+        self.self_voted: bool = False
+        self._poll: Poll = poll
+
+    def _handle_vote_event(self, added: bool) -> None:
```
This should be updating self_voted or whatever we're calling it.

bijij commented on discord/abc.py:
```diff
@@ -1350,6 +1351,7 @@ async def send(
         view: View = ...,
         suppress_embeds: bool = ...,
         silent: bool = ...,
+        poll: Poll = ...,
```
This doesn't seem right?
Unless something changed, you can't send a poll with any other form of "content", so there should be no overloads which accept both Poll and any other kind (e.g. content, view, embeds, attachments)

DA-344 replied:
I've been testing polls and they do can be sent with embeds, views and contents.
![thingy](https://github.com/Rapptz/discord.py/assets/108473820/d5d4eaba-f85c-4426-a4ed-14eec95799fe)

bijij commented on discord/client.py:
```diff
@@ -1834,6 +1836,28 @@ async def wait_for(
+        event: Literal['poll_vote_add', 'poll_vote_remove'],
+        /,
+        *,
+        check: Optional[Callable[[Union[User, Member], Message, PollAnswer], bool]] = None,
+        timeout: Optional[float] = None,
+    ) -> Tuple[Union[User, Member], PollAnswer]:
```
The check type and the return tuple don't match, which one is correct?

bijij commented on discord/ext/commands/context.py:
```diff
@@ -641,6 +642,7 @@ async def reply(
         suppress_embeds: bool = ...,
         ephemeral: bool = ...,
         silent: bool = ...,
+        poll: Poll = ...,
```
See my comment in abc.py

bijij commented on discord/http.py:
```diff
@@ -256,6 +259,13 @@ def handle_message_parameters(
         payload.update(channel_payload)
 
+    if poll is not MISSING:
+        if len(poll) == 0 or len(poll) > 10:
+            raise ValueError('Poll must contain between 1 and 10 answers')
+        if poll._hours_duration < 1 or poll._hours_duration > 168:
+            raise ValueError('Polls duration must be between 1 hour and 7 days')
```
I'm going to reiterate this, I'm still not sure we should be doing this.

bijij commented on discord/http.py:
```diff
@@ -2503,6 +2513,43 @@ def edit_application_info(self, *, reason: Optional[str], payload: Any) -> Respo
+        if after:
+            params['after'] = int(after)
+
+        if limit:
```
This is going to make `0` return 25 results instead of erroring, I think this should be an explicit `None` check instead.
```suggestion
        if limit is not None:
```

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,641 @@
+
+        payload: PollMediaPayload = {'text': self.text}
+
+        if self.emoji:
+            if isinstance(self.emoji, Emoji):
+                payload['emoji'] = {'name': self.emoji.name, 'id': self.emoji.id}
+
+            elif isinstance(self.emoji, PartialEmoji):
+                payload['emoji'] = self.emoji.to_dict()
+
+            else:
+                payload['emoji'] = {'name': str(self.emoji)}  # type: ignore
```
```suggestion
        if isinstance(self.emoji, str):
            payload['emoji'] = {"id": None, "name": self.emoji}
        elif self.emoji is not None:
            payload['emoji'] = {"id": self.emoji.id, "name": self.emoji.name}
```

This can be refworked to remove the type-ignore.

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,641 @@
+        """:class:`Poll`: Returns the parent poll of this answer"""
+        return self._poll
+
+    def _to_dict(self) -> PollMediaPayload:
+        data: PollMediaPayload = {'text': self.text}
+
+        if self.emoji is not None:
+            if isinstance(self.emoji, PartialEmoji):
+                data['emoji'] = self.emoji.to_dict()
+            else:
+                data['emoji'] = {'name': str(self.emoji)}  # type: ignore
+
+                if hasattr(self.emoji, 'id'):
+                    data['emoji']['id'] = int(self.emoji.id)  # type: ignore
```
```suggestion
    def _to_dict(self) -> PollAnswerPayload:
        return {
            'poll_media': self.media.to_dict(),
        }
```

Why isn't this returning the full poll answer payload?

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,641 @@
+                break
+
+            limit -= len(users)
+            after = Object(id=int((users[-1])['id']))
```
```suggestion
            after = Object(id=int(users[-1]['id']))
```
stylistic choice, but I think these parens do nothing but make it harder to read.

DA-344 replied:
Yeah that was there from previous changes that I reverted and forgot to remove the parentheses. Should be removed by now.

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,641 @@
+
+            Returns the Poll's question
+
+        .. describe:: len(x)
```
If you're supporting len, you should support iter IMO.

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,641 @@
+        # _from_data, so it should be ``None`` now.
+        self._message: Optional[Message] = None
+        self._state: Optional[ConnectionState] = None
+        self._results: Optional[PollResultPayload] = None
```
Why are you storing a payload?

DA-344 replied:
MB, updated in latest commits.

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,641 @@
+        return self._answers.copy()
+
+    @property
+    def expiry(self) -> Optional[datetime.datetime]:
```
```suggestion
    def expires_at(self) -> Optional[datetime.datetime]:
```

DA-344 replied:
Change applied in latest commits.

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,641 @@
+        fetched message. You should consider accessing this method
+        via :attr:`Message.poll`
+        """
+        if not self._results:
+            return False
+        return self._results['is_finalized']
```
I'm not sure why you're not just storing the state as a bool, storing the result payload seems like a waste.

DA-344 replied:
Yeah, this is now updated in the latest commits.

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,641 @@
+        """:class:`int`: Returns the sum of all the answer votes."""
+        return sum([answer.vote_count for answer in self.answers])
+
+    def is_finalized(self) -> bool:
```
This needs an EN-GB equivalent.

DA-344 replied:
Added its equivalent (``is_finalised``)

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,641 @@
+    ) -> Optional[PollAnswer]:
+        """Returns the answer with the provided ID or ``None`` if not found.
+
+        Note that the ID, as Discord says, it is the index / row where the answer
```
This is an implementation detail you can't rely on it.

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,641 @@
+        data: PollCreatePayload = {
+            'allow_multiselect': self.multiselect,
+            'question': self._question_media.to_dict(),
+            'duration': self._hours_duration,
```
This whole property seems dangerous, you allow people to add new answers to a copy, but if they edit the duration this won't be updated.

DA-344 replied:
Thanks for the comment, this has been updated so the duration is calculated when calling _to_dict.

bijij commented on discord/poll.py:
```diff
@@ -0,0 +1,641 @@
+            'question': self._question_media.to_dict(),
+            'duration': self._hours_duration,
+            'layout_type': self.layout_type.value,
+            'answers': [{'poll_media': answer._to_dict()} for answer in self.answers],
```
```suggestion
            'answers': [answer.to_dict() for answer in self.answers],
```

If you take my above suggestion, this can be simplified.

bijij reviewed (changes requested):
There's still some substantial issues with the Poll model itself, and some questional UX decisions which may lead to issues in the future.

AbstractUmbra commented:
What's the current status of this PR? Are we progressing, as I can see the last few comments from @bijij are unresolved?

DA-344 commented:
> What's the current status of this PR? Are we progressing, as I can see the last few comments from @bijij are unresolved?

I am applying the suggested changes, most of them are already implemented and I forgot to add a comment and resolve the conversations, lol sorry.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+
+    @property
+    def text(self) -> str:
+        """:class:`str`: Returns this answer display text."""
```
```suggestion
        """:class:`str`: Returns this answer's displayed text."""
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+
+    @property
+    def emoji(self) -> Optional[Union[PartialEmoji, Emoji]]:
+        """Optional[:class:`PartialEmoji`]: Returns this answer display emoji, is any."""
```
```suggestion
        """Optional[Union[:class:`Emoji`, :class:`PartialEmoji`]]: Returns this answer's displayed emoji, if any."""
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+    @property
+    def emoji(self) -> Optional[Union[PartialEmoji, Emoji]]:
+        """Optional[:class:`PartialEmoji`]: Returns this answer display emoji, is any."""
+        if isinstance(self.media.emoji, str):
+            return PartialEmoji.from_str(self.media.emoji)
+        return self.media.emoji
```
You're doing this here but this should have been done in the constructor for `PollMedia`.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        return cls(text=data['text'])
+
+
+class PollAnswer:
```
Missing `versionadded`.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+PollMediaEmoji = Union[PartialEmoji, Emoji, str]
+
+
+class PollMedia:
```
This isn't a `NamedTuple`, nor should it be. Either way, this is missing `versionadded`.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        poll: Poll,
+        message: Optional[Message],
+    ) -> Self:
+        poll_media: PollMediaPayload = {'text': text}
+        if emoji:
+            if isinstance(emoji, Emoji):
+                poll_media['emoji'] = {'name': emoji.name}  # type: ignore
+
+                if emoji.id:
+                    poll_media['emoji']['id'] = emoji.id  # type: ignore
+            elif isinstance(emoji, PartialEmoji):
+                poll_media['emoji'] = emoji.to_dict()
+            else:
+                poll_media['emoji'] = {'name': str(emoji)}  # type: ignore
```
A lot of this can be simplified by doing something similar to this:
```suggestion
        poll_media: PollMediaPayload = {'text': text}
        if emoji is not None:
            emoji = PartialEmoji.from_str(emoji) if isinstance(emoji, str) else emoji._as_partial()
            emoji_data = emoji.to_dict()
            # Remove animated key?
            # emoji_data.pop('animated', None)
            poll_media['emoji'] = emoji_data
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+            return PartialEmoji.from_str(self.media.emoji)
+        return self.media.emoji
+
+    @property
+    def vote_count(self) -> int:
+        """:class:`int`: Returns an approximate votes of this answer. If the poll
+        has finalized, it returns the exact amount of votes.
+        """
+        return self._vote_count
```
```suggestion
    @property
    def vote_count(self) -> int:
        """:class:`int`: Returns an approximate count of votes for this answer.
        
        If the poll is finished, the count is exact.
        """
        return self._vote_count
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+    async def voters(
+        self, *, limit: Optional[int] = None, after: Optional[Snowflake] = None
+    ) -> AsyncIterator[Union[User, Member]]:
+        """Returns an :term:`asynchronous iterator` representing the users that have voted to this answer.
```
```suggestion
        """Returns an :term:`asynchronous iterator` representing the users that have voted on this answer.
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        The ``after`` parameter must represent a user
+        and meet the :class:`abc.Snowflake` abc.
+
+        .. note::
+
+            This can only be called when the parent poll has an attached message.
```
```suggestion
        This can only be called when the parent poll was sent to a message.
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        Yields
+        ------
+        Union[:class:`User`, :class:`Member`]
+            The member (if retrievable) or the user that has voted
+            to this poll answer. The case where it can be a :class:`Member`
+            is in a guild message context. Sometimes it can be a :class:`User`
+            if the member has left the guild.
```
```suggestion
            The member (if retrievable) or the user that has voted
            on this poll answer. The case where it can be a :class:`Member`
            is in a guild message context. Sometimes it can be a :class:`User`
            if the member has left the guild or if the member is not cached.
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        """
+
+        if not self._message or not self._state:  # Make type checker happy
+            raise RuntimeError('You cannot fetch users in a non-message-attached poll')
```
```suggestion
            raise ClientException('You cannot fetch users to a poll not sent with a message')
```

Import `ClientException` and document this in `Raises`.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+            users = data['users']
+
+            if len(users) == 0:
+                # No more voters to fetch, terminate process
```
```suggestion
                # No more voters to fetch, terminate loop
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+class Poll:
+    """Represents a message's Poll.
+
+    .. container:: operations
+
+        .. describe:: str(x)
+
+            Returns the Poll's question
+
+        .. describe:: len(x)
+
+            Returns the Poll's answer amount.
+
```
```suggestion
```

None of these operations actually make sense.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+
+    Parameters
+    ----------
+    question: Union[:class:`PollMedia`, :class:`str`]
```
If the parameter type and the attribute types are different then you need a dedicated attribute section as well.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+    Parameters
+    ----------
+    question: Union[:class:`PollMedia`, :class:`str`]
+        The poll's question media. Text can be up to 300 characters.
```
```suggestion
        The poll's displayed question. The text can be up to 300 characters.
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,640 @@
+            At the moment, this *does not* support emojis.
+
+    duration: :class:`datetime.timedelta`
+        The duration of the poll.
```
Why was this ignored?

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        limit: Optional[:class:`int`]
+            The maximum number of results to return.
+            If not provided, returns all the users who
+            voted to this poll answer.
```
```suggestion
            voted on this poll answer.
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        The poll's question media. Text can be up to 300 characters.
+    duration: :class:`datetime.timedelta`
+        The duration of the poll.
+    multiselect: :class:`bool`
```
Rename this attribute to `multiple` a la HTML since it's a better name.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+    duration: :class:`datetime.timedelta`
+        The duration of the poll.
+    multiselect: :class:`bool`
+        Whether users are allowed to select more than one answer. Defaults
+        to ``True``
```
```suggestion
        Whether users are allowed to select more than one answer. 
        Defaults to ``True``.
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        Whether users are allowed to select more than one answer. Defaults
+        to ``True``
+    layout_type: :class:`PollLayoutType`
+        The layout type of the poll. Defaults to :attr:`PollLayoutType.default`
```
```suggestion
        The layout type of the poll. Defaults to :attr:`PollLayoutType.default`.
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,630 @@
+            self._question_media: PollMedia = (
+                question  # At the moment this only supports text, so no need to add emoji support
+            )
+        self._answers: List[PollAnswer] = []
```
Why was this ignored?

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        self._message = message
+
+        if not message.poll:
+            # Something should go very wrong for this if block to be called
```
```suggestion
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        if not message.poll:
+            # Something should go very wrong for this if block to be called
+            return
+        # These attributes are accessed via message.poll as there
+        # is where all the data is stored
```
```suggestion
        # The message's poll contains the more up to date data.
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        }
+        return data
+
+    def __str__(self) -> str:
+        return self._question_media.text
```
```suggestion
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+    def __repr__(self) -> str:
+        return f"<Poll duration={self.duration} question=\"{self.question}\" answers={self.answers}>"
+
+    def __len__(self) -> int:
+        return len(self.answers)
```
```suggestion
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+class PollMedia:
+    """Represents the poll media for a poll item.
+
+    .. container:: operations
+
+        .. describe:: str(x)
+
+            Returns this Poll Media\'s text.
```
```suggestion
```

I don't think this is future proof.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        self.text: str = text
+        self.emoji: Optional[PollMediaEmoji] = emoji
+
+    def __str__(self) -> str:
+        return self.text
```
```suggestion
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        if isinstance(question, str):
+            self._question_media: PollMedia = PollMedia(text=question, emoji=None)
+        else:
+            self._question_media: PollMedia = (
+                question  # At the moment this only supports text, so no need to add emoji support
+            )
```
```suggestion
            # At the moment this only supports text, so no need to add emoji support
            self._question_media: PollMedia = question
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        """:class:`int`: Returns the sum of all the answer votes."""
+        return sum([answer.vote_count for answer in self.answers])
+
+    def is_finalized(self) -> bool:
```
```suggestion
    def is_finalised(self) -> bool:
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        return sum([answer.vote_count for answer in self.answers])
+
+    def is_finalized(self) -> bool:
+        """:class:`bool`: Returns whether the poll has finalized.
```
```suggestion
        """:class:`bool`: Returns whether the poll has finalised.
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        """
+        return self._finalized
+
+    is_finalised = is_finalized
```
```suggestion
    is_finalized = is_finalised
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+    def is_finalized(self) -> bool:
+        """:class:`bool`: Returns whether the poll has finalized.
+
+        It always returns ``False`` for stateless polls.
```
```suggestion
        This always returns ``False`` for stateless polls.
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+
+    is_finalised = is_finalized
+
+    def copy(self) -> Self:
+        """Returns a stateless copy of this poll.
+
+        This is meant to be used when you want to edit a stateful poll.
+
+        Returns
+        -------
+        :class:`Poll`
+            The copy of the poll.
+        """
```
This looks like a partial deep copy. I guess the name might be fine, but it's a bit weird for a method named `copy` to do a (partial) deep copy rather than a shallow copy.

DA-344 replied:
Yeah I thought of it when I created this method, it does a partial deep copy because, as the docstring says, it is meant to return a stateless poll copy when the actual object has a state. I don't know if I explain myself...

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        """
+
+        new = self.__class__(question=self.question, duration=self.duration)
+        # We want to return a stateless copy of the poll, so we should not
+        # override new._answers as our answers may contain a state
+
```
```suggestion

        # We want to return a stateless copy of the poll, so we should not
        # override new._answers as our answers may contain a state
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        """
+
+        if self._message:
+            raise RuntimeError('Cannot append answers to a poll recieved via message.')
```
```suggestion
            raise ClientException('Cannot append answers to a poll that is active')
```

Update docstring too.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        Returns
+        -------
+        :class:`Poll`
+            This poll with the new answer appended.
```
```suggestion
            This poll with the new answer appended. This is to allow fluent-style chaining.
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+    ) -> Optional[PollAnswer]:
+        """Returns the answer with the provided ID or ``None`` if not found.
+
+        Note that the ID, as Discord says, it is the index / row where the answer
+        is located in the poll.
```
```suggestion
```

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+            The answer.
+        """
+
+        return utils.get(self.answers, id=id)
```
You ignored this earlier but `get_` methods **should not** be O(n). Either change it to a dictionary or remove the method entirely.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        """
+
+        if not self._message or not self._state:  # Make type checker happy
+            raise RuntimeError('This poll has no attached message.')
```
Once again, `ClientException`.

Rapptz commented on discord/raw_models.py:
```diff
@@ -503,3 +505,33 @@ def __init__(self, *, data: GuildApplicationCommandPermissions, state: Connectio
+    Attributes
+    ----------
+    user_id: :class:`int`
+        The ID of the user that added their vote.
```
```suggestion
        The ID of the user that added or removed a vote.
```

Rapptz commented on discord/raw_models.py:
```diff
@@ -503,3 +505,33 @@ def __init__(self, *, data: GuildApplicationCommandPermissions, state: Connectio
+    user_id: :class:`int`
+        The ID of the user that added their vote.
+    channel_id: :class:`int`
+        The Channel ID this poll answer was added.
```
```suggestion
        The channel ID where the poll vote action took place.
```

Rapptz commented on discord/raw_models.py:
```diff
@@ -503,3 +505,33 @@ def __init__(self, *, data: GuildApplicationCommandPermissions, state: Connectio
+    channel_id: :class:`int`
+        The Channel ID this poll answer was added.
+    message_id: :class:`int`
+        The Message ID that contains the poll the user added their vote to.
```
```suggestion
        The message ID that contains the poll the user added or removed their vote on.
```

Rapptz commented on discord/raw_models.py:
```diff
@@ -503,3 +505,33 @@ def __init__(self, *, data: GuildApplicationCommandPermissions, state: Connectio
+    message_id: :class:`int`
+        The Message ID that contains the poll the user added their vote to.
+    guild_id: Optional[:class:`int`]
+        The Guild ID the poll is in, or ``None`` if in a GDM or DM channel.
```
```suggestion
       The guild ID where the vote got added or removed, if applicable.
```

Rapptz commented on discord/raw_models.py:
```diff
@@ -503,3 +505,33 @@ def __init__(self, *, data: GuildApplicationCommandPermissions, state: Connectio
+    guild_id: Optional[:class:`int`]
+        The Guild ID the poll is in, or ``None`` if in a GDM or DM channel.
+    answer_id: :class:`int`
+        The poll answer's ID the user voted to.
```
```suggestion
        The poll answer's ID the user voted on.
```

Rapptz commented on discord/webhook/async_.py:
```diff
@@ -1734,6 +1742,15 @@ async def send(
             .. versionadded:: 2.4
 
+        poll: :class:`Poll`
+            The attached poll for this message.
```
```suggestion
            The poll to send with this message.
```

Rapptz commented on discord/webhook/sync.py:
```diff
@@ -979,6 +983,14 @@ def send(
 
             .. versionadded:: 2.2
+        poll: :class:`Poll`
+            The attached poll for this message.
```
```suggestion
            The poll to send with this message.
```

Rapptz commented on docs/api.rst:
```diff
@@ -1442,6 +1442,71 @@ Voice
     :param after: The voice state after the changes.
     :type after: :class:`VoiceState`
 
+Polls
```
These events are alphabetically sorted, P comes before V not after. The English alphabet is `ABCDEFGHIJLKMNOPQRSTUVWXYZ`

Rapptz commented on docs/api.rst:
```diff
@@ -1442,6 +1442,71 @@ Voice
+
+.. function:: on_poll_vote_add(user, answer)
+
+    Called when a :class:`Message`\'s attached :class:`Poll` gets a new vote. If any of ``user`` or ``message``
+    are not cached this event will not be called.
```
```suggestion
    Called when a :class:`Message`\'s attached :class:`Poll` gets a new vote. If the ``user`` or ``message``
    are not cached then this event will not be called.
```

Rapptz commented on docs/api.rst:
```diff
@@ -1442,6 +1442,71 @@ Voice
+
+    .. note::
+
+        If the poll allows multiselect and the user votes for more than one answer
```
```suggestion
        If the poll allows multiple answers and the user votes for more than one answer
```

Rapptz commented on docs/api.rst:
```diff
@@ -1442,6 +1442,71 @@ Voice
+
+    :param user: The user that voted for the message's poll.
+    :type user: Union[:class:`User`, :class:`Member`]
+    :param answer: The answer the user voted to.
```
```suggestion
    :param answer: The answer the user voted on.
```

Rapptz commented on docs/api.rst:
```diff
@@ -1442,6 +1442,71 @@ Voice
+    :param answer: The answer the user voted to.
+    :type answer: :class:`PollAnswer`
+
+.. function:: on_poll_vote_remove(user, answer)
+
+    Called when a :class:`Message`\'s attached :class:`Poll` has lost a vote. If any of ``user`` or ``message``
+    are not cached this event will not be called.
+
+    This requires :attr:`Intents.message_content` and :attr:`Intents.polls` to be enabled.
+
+    .. note::
+
+        If the poll allows multiselect and the user removes more than one vote
+        this event will be called as many times as votes removed.
+
+    .. versionadded:: 2.4
+
+    :param user: The user that removed their vote from the message's poll.
+    :type user: Union[:class:`User`, :class:`Member`]
+    :param answer: The answer the user removed the vote from.
+    :type answer: :class:`PollAnswer`
```
You should collapse this with the one above, similar to the other events.

Rapptz reviewed (changes requested):
Please fix the CI failures.

Rapptz commented on discord/permissions.py:
```diff
@@ -850,6 +866,7 @@ class PermissionOverwrite:
         send_voice_messages: Optional[bool]
         create_expressions: Optional[bool]
         create_events: Optional[bool]
+        send_polls: Optional[bool]
```
The alias needs to be here too.

DA-344 replied:
Added the alias.

DA-344 commented on discord/message.py:
```diff
@@ -1439,6 +1444,28 @@ async def reply(self, content: Optional[str] = None, **kwargs: Any) -> Message:
 
         return await self.channel.send(content, reference=self, **kwargs)
 
+    async def end_poll(self) -> Message:
```
Okay, I have added this.

DA-344 commented on discord/abc.py:
```diff
@@ -1516,6 +1522,10 @@ async def send(
 
             .. versionadded:: 2.2
+        poll: :class:`~discord.Poll`
+            The attached poll for this message.
```
Commited this change.

DA-344 commented on discord/message.py:
```diff
@@ -1439,6 +1444,28 @@ async def reply(self, content: Optional[str] = None, **kwargs: Any) -> Message:
+    async def end_poll(self) -> Message:
+        """|coro|
+
+        Tries to end this message\'s :class:`Poll`.
```
Committed this change

DA-344 commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+
+    def __init__(self, /, text: str, emoji: Optional[PollMediaEmoji] = None) -> None:
+        self.text: str = text
+        self.emoji: Optional[PollMediaEmoji] = emoji
```
Thank you for the clarification. I've updated the code to parse strings to ``PartialEmoji``s.

DA-344 commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+    Attributes
+    ----------
+    text: :class:`str`
+        The displayed text
```
I've commented in the wrong review 🤦‍♂️. Anyways, apologies for the oversight. Added the dot.

DA-344 commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        if isinstance(self.emoji, str):
+            payload['emoji'] = {"id": None, "name": self.emoji}
+        elif self.emoji is not None:
+            payload['emoji'] = {"id": self.emoji.id, "name": self.emoji.name}
```
I've updated the code to use single quotes instead of double quotes, as requested.

DA-344 commented on discord/poll.py:
```diff
@@ -0,0 +1,624 @@
+        The displayed text
+    emoji: Optional[Union[:class:`PartialEmoji`, :class:`Emoji`, :class:`str`]]
+        The attached emoji for this media. This will always be ignored for a poll
+        question media.
```
Thanks for the improvement, committed the change.

Rapptz commented on discord/poll.py:
```diff
@@ -0,0 +1,577 @@
+        """:class:`str`: Returns this poll answer question string."""
+        return self._question_media.text
+
+    @property
+    def emoji(self) -> Optional[PartialEmoji]:
+        """Optional[:class:`PartialEmoji`]: Returns the emoji for this poll's question."""
+
+        return None  # As of now, polls questions don't support emojis
```
```suggestion
```

I don't see the point of this.

DA-344 replied:
Yeah, I thought of it in the last day, but just couldn't find the words to say it. Anyways, it was removed in the latest commits.

Rapptz reviewed (changes requested):
Just one final nit

DA-344 commented on discord/client.py:
```diff
@@ -1839,6 +1841,28 @@ async def wait_for(
 
+    @overload
+    async def wait_for(
+        self: Union[Bot, AutoShardedBot],
+        event: Literal['poll_vote_add', 'poll_vote_remove'],
+        /,
+        *,
+        check: Optional[Callable[[Union[User, Member], PollAnswer], bool]] = None,
+        timeout: Optional[float] = None,
+    ) -> Tuple[Union[User, Member], PollAnswer]:
+        ...
+
+    @overload
+    async def wait_for(
+        self: Union[Bot, AutoShardedBot],
```
Yeah you are right, thanks for letting me know.

Rapptz reviewed (approved):
Thanks!

mwpowellhtx commented:
How far along is this? Been in a state of limbo for what, over a year?

Soheab commented:
> How far along is this? Been in a state of limbo for what, over a year?

Do you not see the purple "Merged" pill near the name?

mwpowellhtx commented:
> > How far along is this? Been in a state of limbo for what, over a year?
> 
> Do you not see the purple "Merged" pill near the name?

I did not, thank you for mentioning. Are there use case examples?
